### ISSUE TITLE
Add Enum.with_index(enum, fun)

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#10547]
The fun receives two arguments, the element and the index. The following signatures must be valid:

  * `with_index(enum, offset)`
  * `with_index(enum, fun)`
  * ~`with_index(enum, offset, fun)`~

**Comment by fertapric** [comment]
Maybe `with_index(enum, offset, fun)` is not needed because it can be easily done by adding the desired offset in the function:

```elixir
Enum.with_index([:a, :b, :c], 1, fn item, index -> {index, item} end)
Enum.with_index([:a, :b, :c], fn item, index -> {index + 1, item} end)
```

`with_index(enum, offset)` is different because there is no function involved (and because we cannot change it without introducing a backwards incompatible change).

**Comment by josevalim** [comment]
That's a good point. I will remove it from the issue.

**Comment by axelson** [comment]
This was implemented in #10552
